src/queue/priority.rs

```diff
@@ -0,0 +1,501 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A priority task queue. Tasks are scheduled based on its priority, tasks with small priority
+//! value will be scheduler earlier than bigger onces.. User should implement The [`TaskPriorityProvider`]
+//! to provide the priority value for each task. The priority value is fetched from the
+//! [`TaskPriorityProvider`] at each time the task is scheduled.
+//!
+//! The task queue requires that the accompanying [`PriorityRunner`] must beused to collect necessary
+//! information.
+
+use std::{
+    sync::{
+        atomic::{AtomicPtr, AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use crossbeam_skiplist::SkipMap;
+use prometheus::local::LocalIntCounter;
+use prometheus::IntCounter;
+
+use crate::metrics::*;
+use crate::pool::{Local, Runner, RunnerBuilder};
+use crate::queue::{
+    multilevel::{TaskLevelManager, FLUSH_LOCAL_THRESHOLD_US, LEVEL_NUM},
+    Extras, Pop, TaskCell,
+};
+
+// a wrapper of u64 with an extra sequence number to avoid duplicate value.
+#[derive(Eq, PartialEq, Ord, PartialOrd)]
+struct MapKey(u64, u64);
+
+/// The injector of a single level work stealing task queue.
+#[derive(Clone)]
+pub struct TaskInjector<T> {
+    queue: Arc<QueueCore<T>>,
+    task_manager: PriorityTaskManager,
+}
+
+fn set_schedule_time<T>(task_cell: &mut T)
+where
+    T: TaskCell,
+{
+    task_cell.mut_extras().schedule_time = Some(Instant::now());
+}
+
+impl<T> TaskInjector<T>
+where
+    T: TaskCell + Send,
+{
+    /// Pushes the task cell to the queue. The schedule time in the extras is
+    /// assigned to be now.
+    pub fn push(&self, mut task_cell: T) {
+        let priority = self.task_manager.prepare_before_push(&mut task_cell);
+        set_schedule_time(&mut task_cell);
+        self.queue.push(task_cell, priority);
+    }
+}
+
+/// The local queue is just a proxy of the global queue.
+pub(crate) struct LocalQueue<T> {
+    queue: Arc<QueueCore<T>>,
+    task_manager: PriorityTaskManager,
+}
+
+impl<T> LocalQueue<T>
+where
+    T: TaskCell + Send,
+{
+    pub(super) fn push(&mut self, mut task_cell: T) {
+        let priority = self.task_manager.prepare_before_push(&mut task_cell);
+        set_schedule_time(&mut task_cell);
+        self.queue.push(task_cell, priority);
+    }
+
+    pub(super) fn pop(&mut self) -> Option<Pop<T>> {
+        self.queue.pop()
+    }
+
+    pub(super) fn has_tasks_or_pull(&mut self) -> bool {
+        !self.queue.is_empty()
+    }
+}
+
+/// A trait used to generate priority value for each task.
+pub trait TaskPriorityProvider: Send + Sync + 'static {
+    /// Return a priority value of this task, all tasks in the priority queue is ordered by this value.
+    fn get_priority(&self, extras: &Extras) -> u64;
+}
+
+///
+#[derive(Clone)]
+struct PriorityTaskManager {
+    level_manager: Arc<TaskLevelManager>,
```

Comment:
With a priority queue, why is there still a level manager?

Reply:
While using the priority queue, we still need to auto demote slow tasks. So the priority manager can use the level information to set different priority value. Or we have to implement similar mechanism to do so.